### ISSUE TITLE
Support different storage engines / flavors in backup & restore

### ISSUE THREAD

**Issue by derekperkins** [vitessio/vitess#4498]
We are looking into migrating some of our workloads to use MyRocks. Our preferred way to approach this would be to set up 1+ replicas per shard to use MyRocks, keeping master on Innodb. I would like to have backups / restores happening for each type independently, or at a minimum, only backup whatever is currently the serving master.

A similar approach would be required to switch between MySQL, Percona, MariaDB or even to try out upgraded versions like 8.0 that aren't filesystem compatible. (upgrades would work, but downgrades wouldn't)

Maybe there is a separate approach that would work better. A lot of these thoughts are based on a few Slack conversations with @sjmudd 

> With the comments you made earlier about moving to 8.0 this does bring up the issue of which version is backed up. I often run for normal MySQL multiple major versions in the same replication chain. So knowing the backup version is quite important.  It’s not clear to me how Vitess handles versions. I have seen that it does an auto mysql_upgrade on restore which is good but I wonder if it may need to be version aware to either not backup 8.0 when you run 5.7 plus 8.0 or maybe to backup both versions. I think right now that this is simply ignored.  So looks like something that may need _some_ attention at some point if there are version incompatibilities on all the tablets running in a keyspace.

https://vitess.slack.com/archives/C0PQY0PTK/p1540794850056400

**Comment by sjmudd** [comment]
* I think that Vitess should be version aware
* If it is you can then come up with a policy for doing backups which match the required policy.  

Apart from the suggested options of: master version, all versions, specific version (naming the specific versions) be aware that I've seen due to bugs and other things that there may be an interest in also taking backups of all different minor versions too. Not all minor versions are _exactly_ the same so while one minor version may be ok having a copy of all of them may be better.

* with backups you also have restores and a policy would be needed there too, though generally that would be to restore from the latest version that is not greater than the MySQL version being restored to.

I don't have details at hand but I know that I've been running on the same replication chain 3 different major versions of MySQL and 2 of MariaDB. With larger installations this happens and it may also happen if you want to test and evaluate newer versions of MariaDB or MySQL which are not GA yet.

If adding a lot of extra logic to Vitess seems complicated perhaps delegating this to a separate daemon which can be aware of the running MySQL servers (and their versions), the available backups, and the version to restore to, and this would allow the process to indicate which backups to take or which backup to restore when building a new image.

**Comment by derekperkins** [comment]
I agree, Vitess should definitely be version aware. I would also like to add `ENGINE` to the metadata. Alternatively, maybe we just add another open field that is used for comparison, something like `-backup-compatibility-string=myrocks`, and then people could differentiate backups however they want.

> I know that I've been running on the same replication chain 3 different major versions of MySQL and 2 of MariaDB

Wow.
